crates/locutus-runtime/src/versioning.rs

```diff
@@ -0,0 +1,167 @@
+use std::fmt;
+use std::fmt::{Display, Formatter};
+use std::fs::File;
+use std::io::{Cursor, Read};
+use std::path::Path;
+use std::sync::Arc;
+
+use byteorder::{BigEndian, ReadBytesExt};
+use locutus_stdlib::contract_interface::ContractKey;
+use semver::Version;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+
+use crate::WrappedContract;
+use locutus_stdlib::prelude::{ContractCode, Parameters, TryFromTsStd, WsApiError};
+
+/// Wrapper that allows contract versioning. This enum maintains the types of contracts that are
+/// allowed and their corresponding version.
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq)]
+pub enum ContractContainer {
+    Wasm(WasmAPIVersion),
+}
+
+impl ContractContainer {
+    /// Return the `ContractContainer` content from the specified path as as `Vec<u8>`.
+    pub(crate) fn get_contract_data_from_fs(path: &Path) -> Result<Vec<u8>, std::io::Error> {
+        let mut contract_file = File::open(path)?;
+        let mut contract_data = if let Ok(md) = contract_file.metadata() {
+            Vec::with_capacity(md.len() as usize)
+        } else {
+            Vec::new()
+        };
+        contract_file.read_to_end(&mut contract_data)?;
+        Ok(contract_data)
+    }
+
+    /// Return the `ContractKey` from the specific contract version.
+    pub fn get_key(&self) -> ContractKey {
+        match self {
+            Self::Wasm(WasmAPIVersion::V0_0_1(contract_v1)) => contract_v1.key().clone(),
+        }
+    }
+
+    /// Return the `Parameters` from the specific contract version.
+    pub fn get_params(&self) -> Parameters<'static> {
+        match self {
+            Self::Wasm(WasmAPIVersion::V0_0_1(contract_v1)) => contract_v1.params().clone(),
+        }
+    }
+
+    /// Return the contract code from the specific contract version as `Vec<u8>`.
+    pub fn get_data(&self) -> Vec<u8> {
+        match self {
+            Self::Wasm(WasmAPIVersion::V0_0_1(contract_v1)) => {
+                contract_v1.clone().try_into().unwrap()
+            }
+        }
+    }
+}
+
+impl Display for ContractContainer {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
+            ContractContainer::Wasm(wasm_version) => {
+                write!(f, "Wasm container {}", wasm_version)
+            }
+        }
+    }
+}
+
+impl<'a> TryFrom<(&'a Path, Parameters<'static>)> for ContractContainer {
+    type Error = WsApiError;
```

Comment:
this ain't really a `WsApiError`, I think we can either use std::io::Error here or a custom other error type.

The kind of errors that are happening here are IO errors OR deserializing errors (which we can fit as a IO error probably, `IncorrectValue`), so I think we could use IO error for now and log the error in case there is a deserialization error

Reply:
Used std::io::ErrorKind::InvalidData instead